From da6fbac4cd604de9570bfd19f0745241a73167a4 Mon Sep 17 00:00:00 2001
From: someone5678 <someone5678@users.noreply.github.com>
Date: Wed, 18 Oct 2023 21:23:42 +0900
Subject: [PATCH 2/2] Settings: Make all user apps cloneable

* Filter system apps and overlays
* Include pre-defined apps in com.android.internal.R.array.cloneable_apps

Change-Id: Ia08a1095cd2aec9c6d77d055f1344e57225a44d5
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../AppStateClonedAppsBridge.java             |  5 ++--
 .../ClonedAppsPreferenceController.java       |  6 ++--
 .../android/settings/custom/utils/AppUtils.kt | 28 +++++++++++++++++++
 3 files changed, 34 insertions(+), 5 deletions(-)
 create mode 100644 src/com/android/settings/custom/utils/AppUtils.kt

diff --git a/src/com/android/settings/applications/AppStateClonedAppsBridge.java b/src/com/android/settings/applications/AppStateClonedAppsBridge.java
index 719023d10e2..4b1ef67fbeb 100644
--- a/src/com/android/settings/applications/AppStateClonedAppsBridge.java
+++ b/src/com/android/settings/applications/AppStateClonedAppsBridge.java
@@ -24,6 +24,7 @@ import static com.android.settingslib.applications.ApplicationsState.AppFilter;
 import android.content.Context;
 import android.util.Log;
 
+import com.android.settings.custom.utils.AppUtils;
 import com.android.settings.Utils;
 import com.android.settingslib.applications.ApplicationsState;
 
@@ -42,13 +43,13 @@ public class AppStateClonedAppsBridge extends AppStateBaseBridge{
     private final List<String> mAllowedApps;
     private List<String> mCloneProfileApps = new ArrayList<>();
     private int mCloneUserId;
+    private AppUtils appUtils = new AppUtils();
 
     public AppStateClonedAppsBridge(Context context, ApplicationsState appState,
             Callback callback) {
         super(appState, callback);
         mContext = context;
-        mAllowedApps = Arrays.asList(mContext.getResources()
-                .getStringArray(com.android.internal.R.array.cloneable_apps));
+        mAllowedApps = appUtils.getCloneableAppListStr(mContext);
     }
 
     @Override
diff --git a/src/com/android/settings/applications/ClonedAppsPreferenceController.java b/src/com/android/settings/applications/ClonedAppsPreferenceController.java
index d134ceccd88..38f6dcc80f0 100644
--- a/src/com/android/settings/applications/ClonedAppsPreferenceController.java
+++ b/src/com/android/settings/applications/ClonedAppsPreferenceController.java
@@ -30,6 +30,7 @@ import androidx.lifecycle.OnLifecycleEvent;
 import androidx.preference.Preference;
 import androidx.preference.PreferenceScreen;
 
+import com.android.settings.custom.utils.AppUtils;
 import com.android.settings.R;
 import com.android.settings.Utils;
 import com.android.settings.core.BasePreferenceController;
@@ -45,6 +46,7 @@ public class ClonedAppsPreferenceController extends BasePreferenceController
         implements LifecycleObserver {
     private Preference mPreference;
     private Context mContext;
+    private AppUtils appUtils = new AppUtils();
 
     public ClonedAppsPreferenceController(Context context, String preferenceKey) {
         super(context, preferenceKey);
@@ -81,9 +83,7 @@ public class ClonedAppsPreferenceController extends BasePreferenceController
             @Override
             protected Integer[] doInBackground(Void... unused) {
                 // Get list of allowlisted cloneable apps.
-                List<String> cloneableApps = Arrays.asList(
-                        mContext.getResources().getStringArray(
-                                com.android.internal.R.array.cloneable_apps));
+                List<String> cloneableApps = appUtils.getCloneableAppListStr(mContext);
                 List<String> primaryUserApps = mContext.getPackageManager()
                         .getInstalledPackagesAsUser(/* flags*/ 0, UserHandle.myUserId()).stream()
                         .map(x -> x.packageName).toList();
diff --git a/src/com/android/settings/custom/utils/AppUtils.kt b/src/com/android/settings/custom/utils/AppUtils.kt
new file mode 100644
index 00000000000..e08130feac7
--- /dev/null
+++ b/src/com/android/settings/custom/utils/AppUtils.kt
@@ -0,0 +1,28 @@
+package com.android.settings.custom.utils
+
+import android.content.Context
+import android.content.pm.PackageInfo
+import android.os.UserHandle
+
+class AppUtils {
+    public fun getCloneableAppList(context: Context): List<PackageInfo> {
+        var packageList: List<PackageInfo> =
+            context.getPackageManager().getInstalledPackagesAsUser(/* flags*/ 0, UserHandle.myUserId())
+
+        var filteredList: List<PackageInfo> = packageList.filter {
+                context.getResources().getStringArray(
+                    com.android.internal.R.array.cloneable_apps)
+                        .asList().contains(it.applicationInfo.packageName)
+                || (!it.applicationInfo.isSystemApp()
+                    && !it.applicationInfo.isResourceOverlay())
+            }
+    
+        return filteredList
+    }
+
+    public fun getCloneableAppListStr(context: Context): List<String> {
+        return getCloneableAppList(context).map {
+                x -> x.packageName
+            }.toList()
+    }
+}
-- 
2.25.1

